File: src-tauri/crates/app-ui/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

pub const VERSION: u32 = 1;
pub const MAX_BODY: usize = 64 * 1024;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", deny_unknown_fields)]
pub enum Workspace {
    Global {},
    Session {
        #[serde(rename = "sessionId")]
        session_id: String,
    },
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Target {
    pub instance_id: String,
    pub window_id: String,
    pub workspace: Workspace,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Request {
    pub protocol_version: u32,
    pub request_id: String,
    pub command: String,
    pub target: Target,
    #[serde(default = "empty_params")]
    pub params: Value,
    #[serde(default)]
    pub reveal: bool,
    #[serde(default = "default_timeout")]
    pub timeout_ms: u64,
}
fn empty_params() -> Value {
    json!({})
}
fn default_timeout() -> u64 {
    10_000
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Response {
    pub protocol_version: u32,
    pub request_id: String,
    pub status: Status,
    pub target: Target,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<Failure>,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Status {
    Applied,
    Failed,
    Unknown,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Failure {
    pub code: String,
    pub message: String,
}
impl Response {
    pub fn error(request: &Request, code: &str, message: &str, unknown: bool) -> Self {
        Self {
            protocol_version: VERSION,
            request_id: request.request_id.clone(),
            target: request.target.clone(),
            status: if unknown {
                Status::Unknown
            } else {
                Status::Failed
            },
            result: None,
            error: Some(Failure {
                code: code.into(),
                message: message.into(),
            }),
        }
    }
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Dispatch {
    pub generation: String,
    pub request: Request,
}

File: src/services/uiCommands/protocol.ts
import { z } from "zod";

export const workspaceSchema = z.discriminatedUnion("kind", [
  z.strictObject({ kind: z.literal("global") }),
  z.strictObject({
    kind: z.literal("session"),
    sessionId: z.string().min(1).max(512),
  }),
]);
export const requestSchema = z.strictObject({
  protocolVersion: z.literal(1),
  requestId: z.string().min(1).max(128),
  command: z.string(),
  target: z.strictObject({
    instanceId: z.string(),
    windowId: z.literal("main"),
    workspace: workspaceSchema,
  }),
  params: z.record(z.string(), z.unknown()),
  reveal: z.boolean(),
  timeoutMs: z.number().int().min(1).max(30000),
});
export type UiRequest = z.infer<typeof requestSchema>;
// The protocol's workspace, which is narrower than the app's
// `WorkstationWorkspaceKey`: there is no `directory` target on the wire.
export type UiWorkspace = UiRequest["target"]["workspace"];
export interface UiResponse {
  protocolVersion: 1;
  requestId: string;
  target: UiRequest["target"];
  status: "applied" | "failed" | "unknown";
  result?: unknown;
  error?: { code: string; message: string };
}
export function failure(
  request: UiRequest,
  code: string,
  message: string
): UiResponse {
  return {
    protocolVersion: 1,
    requestId: request.requestId,
    target: request.target,
    status: "failed",
    error: { code, message },
  };
}
